### ISSUE TITLE
Disadvantages of the using Unicode strings in ktl::exception

### ISSUE THREAD

**Issue by DymOK93** [DymOK93/KTL#139]

**Comment by DymOK93** [comment]
Yes... This is a mine very serious architectural mistake. Fixing it will break some of the existing code.

According to the [DbgPrint documentation](https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/wdm/nf-wdm-dbgprint), Unicode strings can only be used with IRQL = PASSIVE_LEVEL. I thought the limitation was "<= DISPATCH_LEVEL" and therefore ktl::exception and assert use Unicode strings. Now I am faced with the problem of the impossibility of correct debug output in the production mini-filter driver: during debugging I got a BSOD because the callback function was called with IRQL = APC_LEVEL. 

@GeorgyFirsov, what do you think: it's worth it to replace Unicode strings with ANSI strings in exceptions and assert  (at the same time, however, fixing tons of compilation errors in the existing code)?

**Comment by GeorgyFirsov** [comment]
I think, `what` method of an exception object should return a value of `const char*` type according to C++ standard. But it might be useful to keep `const wchar_t*`-typed description by introducing a new method `ktl::exception::description()`, that can look like follows:
```cpp
const char* description() {
  static string ansi_description = w2a(what()); // Just a concept
  return ansi_description.c_str();
}
```

Fixing assertions might be difficult, but it makes sense before the first major release. 

Furthermore, I consider, that the library must have the most relaxed restrictions available, which means providing debug info with the most possible for `DbgPrint` IRQL level.

**Comment by DymOK93** [comment]
> 
> 
> `what` could be rewritten to look as follows:

Unfortunately, such an implementation of the `what()` is impossible - the kernel simply does not have an API for `w2a()` compatible with IRQL > PASSIVE_LEVEL

**Comment by DymOK93** [comment]
It's required to provide only constructors for `const_char*` __or__ for both `const wchar_t*` and `const_char*`
The second case makes it possible to throw an exception if conversion is not allowed due to too high IRQL, but the idea is interesting - an approach to `boost::exception`.

**Comment by GeorgyFirsov** [comment]
[`RtlUnicodeToMultiByteN`](https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/ntifs/nf-ntifs-rtlunicodetomultibyten) works with IRQL < DISPATCH_LEVEL (i.e. PASSIVE and APC). Better than nothing :) But DISPATCH_LEVEL and above still stay uncovered.

Speaking about general architecture of library it makes sense to have `const char*` return type in `what` (and as mentioned consequently in internal implementations) up to IRQL = DIRQL and ability to get Unicode description even with several IRQLs.

**Comment by DymOK93** [comment]
In my opinion, there are three possible solutions: 
> Storing `const char*` in the internal storage with converting from Unicode/to Unicode if necessary
> Storing both `const wchar_t*` (optional `description`) and `const_char*` (`what`)
> Storing `const_char*` (`what`) and pointer to the `error_info_base`, where `error_info_base` is the base class for the `boost`-like class `template <class Tag, class Ty> struct error_info` providing `const Ty & get_description()` and other interfaces if needed.
The `error_info` must be allocated by the user himself.

**Comment by GeorgyFirsov** [comment]
> Storing `const char*` in the internal storage with converting from Unicode/to Unicode if necessary

It is the most appropriate (at least for now) solution. It doesn't break "allocation-less" throwing of an exception and is quite easy to use:
```cpp
KdPrint((__FUNCTION__ ": [ERROR] %s", exc.what()));
```

If someone really want Unicode string, they still can obtain it if necessary:
```cpp
DRV_TRACE(__FUNCTIONW__ L": [ERROR] %s", exc.description());
```

One of advantages of such approach is that an exception doesn't store potentially unused information.

> Storing both `const wchar_t*` (optional `description`) and `const char*` (`what`)

Well, it stores potentially unused info. And looks a bit weird :smile:

> Storing `const char*` (`what`) and a pointer to the array of `error_info_base`...

As for me it looks scary. Potentially it will be more flexible, but firstly I need to look at implementation, but for now it is strange for me (as a user) to allocate memory by _myself_ to get an exception description.

----

As a discussion question. In the first and the third cases some extra memory must be allocated, and the required amount of it is in general unknown. Hence this memory must be allocated in heap.

Is it actually a good idea to attempt to allocate some memory in case of exception in kernel mode?

**Comment by DymOK93** [comment]
> Is it actually a good idea to attempt to allocate some memory in case of an exception in kernel mode?
Well, by default no memory should be dynamically allocated. 

95% of real exception throws look like ktl::throw_exception <exception_type>("info", constexpr_message_tag {})
It might be more correct to rename the `constexpr_message_tag` to `persistent_message_tag`, but this is not the point at present.

In some rare cases, it's to have to store runtime information in an exception. And for me, it was always Unicode strings :)

**Comment by GeorgyFirsov** [comment]
> Well, by default no memory should be dynamically allocated.

So the only two alternatives we have, are:
1) Store `const wchar_t*` and `const char*` simultaneously
2) Provide access to `const char*` description only. 

Just because the library must not encourage bad practices. 

If someone's need Unicode description, they can convert it manually. Moreover, I think, _every_ existing large project has it's own strings conversion functions

**Comment by DymOK93** [comment]
I suggest the following public interface for [`exception_base`] (https://github.com/DymOK93/KTL/blob/master/runtime/include/exception_base.h#L15):

```c++
    class exception_base {
    public:
        exception_base(const char* msg); // Copies the message to the allocated buffer 
        exception_base(const char* msg, size_t length); // Copies the message to the allocated buffer 
        
        exception_base(const wchar_t* msg); // Converts the message to the Multibyte and copies to the allocated buffer 
        exception_base(const wchar_t* msg, size_t length); // Converts the message to the Multibyte and copies to the allocated buffer 

        template <size_t N>
        explicit constexpr exception_base(const char(&msg)[N]); // Stores directly a pointer to a string literal 
        
        template <size_t N>
        explicit constexpr exception_base(const char(&msg)[N],
                                          constexpr_message_tag); // Stores directly a pointer to a string literal 

        constexpr exception_base(const char* msg, constexpr_message_tag) noexcept; // Stores directly a pointer to a string literal 

        exception_base(const exception_base& other);
        exception_base& operator=(const exception_base& other);
        virtual ~exception_base();
  /* other methods */
];
```

**Comment by GeorgyFirsov** [comment]
> Copies the message to the allocated buffer

Allocated by ...? Sometimes it may be impossible, and constructor should not even try

**Comment by DymOK93** [comment]
>
> Allocated by ...? Sometimes it may be impossible, and constructor should not even try
We cannot assume that the runtime string will live to `catch` handler. In addition, it's required to provide protection against attempts at allocations to DIRQL.

**Comment by GeorgyFirsov** [comment]
I have a bit weird idea, that solves the problem of dynamic allocation in such situations, but requires introducing explicit initialization and deinitialization of KTL runtime.

Runtime can reserve a couple of pages. But... Using such heavy approach for just Unicode error descriptions looks like an over-kill :)

**Comment by DymOK93** [comment]
Not a bad idea, in my opinion. Then exceptions _almost never_ require allocation. To do this, it's possible to use can use `preload_initializer` appeared in KTL Runtime :)

**Comment by DymOK93** [comment]
I think I'll provide an implementation within a couple of days. Thank you!
